pkg/kubelet/kubelet.go

```diff
@@ -2052,24 +2052,30 @@ func (kl *Kubelet) canAdmitPod(pods []*api.Pod, pod *api.Pod) (bool, string, str
 	}
 	nodeInfo := schedulercache.NewNodeInfo(pods...)
 	nodeInfo.SetNode(node)
-	fit, err := predicates.GeneralPredicates(pod, nil, nodeInfo)
-	if !fit {
-		if re, ok := err.(*predicates.PredicateFailureError); ok {
-			reason := re.PredicateName
-			message := re.Error()
-			glog.V(2).Infof("Predicate failed on Pod: %v, for reason: %v", format.Pod(pod), message)
-			return fit, reason, message
-		}
-		if re, ok := err.(*predicates.InsufficientResourceError); ok {
-			reason := fmt.Sprintf("OutOf%s", re.ResourceName)
-			message := re.Error()
-			glog.V(2).Infof("Predicate failed on Pod: %v, for reason: %v", format.Pod(pod), message)
-			return fit, reason, message
-		}
-		reason := "UnexpectedPredicateFailureType"
+	fit, reasons, err := predicates.GeneralPredicates(pod, nil, nodeInfo)
+	if err != nil {
 		message := fmt.Sprintf("GeneralPredicates failed due to %v, which is unexpected.", err)
 		glog.Warningf("Failed to admit pod %v - %s", format.Pod(pod), message)
-		return fit, reason, message
+		return fit, "UnexpectedError", message
+	}
+	if !fit {
+		// If there are failed predicates, we only return the first one as a reason.
+		if len(reasons) > 0 {
+			r := reasons[0]
+			if re, ok := r.(*predicates.PredicateFailureError); ok {
+				message := re.Error()
+				glog.V(2).Infof("Predicate failed on Pod: %v, for reason: %v", format.Pod(pod), message)
+				return fit, re.PredicateName, message
+			} else if re, ok := r.(*predicates.InsufficientResourceError); ok {
+				message := re.Error()
+				glog.V(2).Infof("Predicate failed on Pod: %v, for reason: %v", format.Pod(pod), message)
+				return fit, fmt.Sprintf("OutOf%s", re.ResourceName), message
+			} else {
+				message := fmt.Sprintf("GeneralPredicates failed due to %v, which is unexpected.", r)
+				glog.Warningf("Failed to admit pod %v - %s", format.Pod(pod), message)
+				return fit, "UnexpectedPredicateFailureType", message
+			}
+		}
```

Comment:
This is broken - if fit is false, you should always return no matter if we have some reasons or not.